### DISCUSSION TITLE
Problem with `git push`.

### DISCUSSION THREAD

**Discussion by LiamM32** [julien-boudry/Condorcet#158] [Collaboration]
Hello. I'm in a bad situation right now with git. I want to push my latest commits for the `Experimental/Schulze-STV` branch of Condorcet. This included a merge commit and more. My branch should be up to date, as I did `git pull` earlier, & the latest commit on the remote branch is from me. But I ran into some problems, and I think I messed things up trying to solve it.

I did `git push origin`. It output:
```
fatal: You are not currently on a branch.
To push the history leading to the current (detached HEAD)
state now, use

    git push origin HEAD:<name-of-remote-branch>
```
I wasn't sure why this is, as my local branch is called `Experimental/Schulze-STV`, just like the remote one. I have pushed from it before.

I tried to find out how to set my branch to the remote branch. I found a page on Stack Overflow that said that I should enter `git checkout Experimental/Schulze-STV`. Here was the output:
```
Warning: you are leaving 1 commit behind, not connected to
any of your branches:

  e96c793c More work on Schulze-STV. It is now closer to outputing the correct result.

If you want to keep it by creating a new branch, this may be a good time
to do so with:

 git branch <new-branch-name> e96c793c

Switched to branch 'Experimental/Schulze-STV'
Your branch is ahead of 'origin/master' by 32 commits.
  (use "git push" to publish your local commits)
```
Oh no! I didn't expect it to overwrite my current local branch with another of the same name, but without the latest commits.

I found [another page on Stack Overflow](https://stackoverflow.com/questions/14365736/warning-you-are-leaving-1-commit-behind-not-connected-to-any-of-your-branches) that said I should do `git cherry-pick e96c793c` (the name of my most recent commit) in this case. But it turns out it didn't do what I wanted. I want to just get all my commits back without making a new one, not to do a merge commit with integrating _some_ changes to the current local branch. So it gave me some edits to my files with the HEAD vs commit e96c793c on various lines for me to decide which ones I wanted. But I already did this earlier, so I shouldn't have to do this again.

So I tried doing `git branch Schulze-STV e96c793c` to see if I can get back to where I was before making that mistake. But it didn't. I did `git log` in that branch, & I don't see my latest commits.

Fortunately, I can still see my latest commits in `git reflog`. But I don't know how to get them back.

I went back to my branch `Experimental/Schulze-STV` to do a `git reset --hard` & get rid of those merge stuff in the files.

Would you be able to quickly give me guidance on how to fix this? You would need to respond quickly, as I'm going away tomorrow.

**Comment by LiamM32** [comment]
@julien-boudry Here are the first few lines of my `git reflog` output:
```
d417485b (HEAD -> Experimental/Schulze-STV) HEAD@{0}: checkout: moving from Experimental/Schulze-STV to Experimental/Schulze-STV
d417485b (HEAD -> Experimental/Schulze-STV) HEAD@{1}: reset: moving to HEAD
d417485b (HEAD -> Experimental/Schulze-STV) HEAD@{2}: reset: moving to HEAD
d417485b (HEAD -> Experimental/Schulze-STV) HEAD@{3}: checkout: moving from e96c793c389ac7c5e99136649a19f02564c503a3 to Experimental/Schulze-STV
e96c793c (Schulze-STV) HEAD@{4}: commit: More work on Schulze-STV. It is now closer to outputing the correct result.
18354294 (origin/Experimental/Schulze-STV) HEAD@{5}: commit: Massive optimization to Schulze_STV.
ea9ecb7d HEAD@{6}: pull origin Experimental/Schulze-STV --rebase (start): checkout ea9ecb7d26d6c9c3de5bbd0cb8c6a1a00c77b5a5
d417485b (HEAD -> Experimental/Schulze-STV) HEAD@{7}: commit: Massive optimization to Schulze_STV.
a468febb HEAD@{8}: Branch: renamed refs/heads/Schulze-STV to refs/heads/Experimental/Schulze-STV
a468febb HEAD@{10}: commit: Undid changes to Schulze_Core.
246195f2 HEAD@{11}: commit: More progress on Schulze-STV. Numerous errors & bugs have been eliminated.
f26d5eaa HEAD@{12}: commit: Some simplification of Schulze STV. Still has errors.
e61dc4ba HEAD@{13}: checkout: moving from dev-4.5 to Schulze-STV
```

**Comment by julien-boudry** [comment]
1. Create a local backup branch, just in case. `git checkout -b backup_something_branch`
2. Commit "Massive optimization to Schulze_STV" (183542941c12839554bf29b0d4dbd1c31a25f245) can't be lost, I see it on remote. And you don't have the right to push git force in this branch. So it's safe.
3. Come back tothe Schulze-Stv branch. Make a git reset hard to an older commit, then a git pull.
4. If something missing, merge your backup branch to this branch, then git push.

At the end, will make some rebasing before merging. So if the history is actually ugly, it's not matter.

**Comment by LiamM32** [comment]
Unfortunate that you couldn't get back to me, but I understand that it is very late in your time zone. I had to figure it out for myself. Edit: Sorry. I did this before you replied. I figured that you must have gone to sleep for the night.

I did `git checkout e96c793c --force`. It seemed to have gotten me back to where I wanted, except with a detached HEAD once again. I did `git push origin HEAD:Experimental/Schulze-STV` and the push was successful.

I can't find any writing that gives a serious answer to how to reattach the HEAD. The best I can find was to do `git checkout Experimental/Schulze-STV`. Here was the output:
```
Previous HEAD position was e96c793c More work on Schulze-STV. It is now closer to outputing the correct result.
Switched to branch 'Experimental/Schulze-STV'
Your branch is ahead of 'origin/master' by 32 commits.
  (use "git push" to publish your local commits)
```
What!? 32 commits? What happened? Did it duplicate them? I don't know where I even begin to resolve this one.

All I can think of is to delete this local branch, & then recreate it from the remote branch.

**Reply by julien-boudry** [reply]
Don't worry ;)

32 commits from master (on this project, it's similar to stable), not from `dev-4.5`.

Yes, if you checkout on a commit (and not to a branch), then you are not on a branch.
Instead checkout to the branch and reset the branch to commit.

Else, yes, maybe try to recreate it.

**Reply by julien-boudry** [reply]
Currently, I have that locally. So your commit is never lost.
![image](https://github.com/julien-boudry/Condorcet/assets/4020317/2978418f-e44b-4155-919e-2203d110bc02)

**Comment by LiamM32** [comment]
So everything appears to be fine now; mostly. I did `git checkout Experimental/Schulze-STV`.

But there's one small thing that's off. I originally placed `Schulze_proportional_prefilter.php` in `src/Algo/Methods/Schulze/`, but then I `git mv` moved it to `src/Algo/Tools` before the latest commit. In your remote repository, it's in this correct location. But in my local branch, it's now back where it was before. Doing `git diff` & `git status` doesn't show this difference, yet if I move it to where it's supposed to be, `git diff` treats it as a change.

**Reply by LiamM32** [reply]
Moving it back & doing `git mv` fixed this, but it's strange behaviour.

**Reply by LiamM32** [reply]
So I noticed some other oddities, like `Condorcet.php` was the old version (despite not showing on `git diff`). I just deleted my local branch and did `git checkout -b Experimental/Schulze-STV origin/Experimental/Schulze-STV`. It looks like everything is good now.
